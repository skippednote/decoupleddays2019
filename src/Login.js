import React, { useEffect, useState } from "react";
import { navigate } from "@reach/router";

function Login() {
  const [token, setToken] = useState(null);
  useEffect(() => {
    window.localStorage.setItem("token", token);
  }, [token]);
  
  const submit = e => {
    e.preventDefault();
    const { username, password } = e.target.elements;
    setToken("Basic " + btoa(`${username.value}:${password.value}`));
    navigate("/");
  };
  return (
    <form onSubmit={submit}>
      <input type="text" placeholder="username" name="username" />
      <input type="password" placeholder="password" name="password" />
      <button type="submit">Login</button>
    </form>
  );
}

export { Login };
