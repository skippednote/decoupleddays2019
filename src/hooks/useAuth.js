import { useState, useContext, useEffect, createContext } from "react";

export const AuthContext = createContext({});

function useAuth() {
  const [token, setToken] = useState(window.localStorage.getItem("token"));

  useEffect(() => {
    if (token) {
      window.localStorage.setItem("token", token);
    } else {
      window.localStorage.removeItem("token");
    }
  }, [token]);

  return { token, setToken };
}

export const useAuthContext = () => useContext(AuthContext);

export { useAuth };
