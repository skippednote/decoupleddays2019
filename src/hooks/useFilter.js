import { useState, useEffect } from "react";
import { getDate } from "../utils/getDate";

export function useFilter(sessions) {
  const [filtered, setFiltered] = useState(sessions);
  const [filter, setFilter] = useState({
    category: "All",
    date: "All",
    input: ""
  });

  useEffect(() => {
    const filteredSession = sessions.filter(session => {
      const { date: sessionDate } = getDate(session.field_date);
      return (
        (filter.category === "All" ||
          session.field_category.name === filter.category) &&
        (filter.date === "All" || sessionDate === filter.date) &&
        (filter.input === "" || session.title.indexOf(filter.input) > -1)
      );
    });

    setFiltered(filteredSession);
  }, [sessions, filter]);

  return { filter, setFilter, filtered };
}
