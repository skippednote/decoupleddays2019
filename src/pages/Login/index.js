import React, { useState, lazy, Suspense } from "react";
import { useAuthContext } from "../../hooks/useAuth";
const LoginForm = lazy(() => import("../../components/LoginForm"));

function Login({ history }) {
  const [status, setStatus] = useState({ error: false, success: false });
  const { token, setToken } = useAuthContext();
  if (token) {
    history.push("/");
    return null;
  }

  const submitHandler = e => {
    e.preventDefault();
    const { username, password } = e.target.elements;
    if (username.value === "" || password.value === "") {
      setStatus({
        error: true,
        success: false
      });
      return;
    }
    const token = "Basic " + btoa(`${username.value}:${password.value}`);
    setToken(token);
    setStatus({
      error: false,
      success: true
    });
    history.push("/");
  };

  return (
    <Suspense fallback={<div>Loading component...</div>}>
      <LoginForm submitHandler={submitHandler} status={status} />
    </Suspense>
  );
}

export default Login;
