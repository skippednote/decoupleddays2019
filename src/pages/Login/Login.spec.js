import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Login from ".";
import { AuthContext } from "../../hooks/useAuth";

const historyMock = { push: jest.fn() };
const setTokenMock = jest.fn();

afterEach(() => {
  setTokenMock.mockClear();
  historyMock.push.mockClear();
});

it("renders the component", async () => {
  const { container, getByPlaceholderText } = render(
    <MemoryRouter>
      <Login />
    </MemoryRouter>
  );
  expect(container).toMatchSnapshot();

  await wait(() =>
    expect(getByPlaceholderText("username")).toBeInTheDocument()
  );
  expect(container).toMatchSnapshot();
});

it("redirects to home if logged in", () => {
  render(
    <MemoryRouter>
      <AuthContext.Provider value={{ setToken: setTokenMock, token: "yo" }}>
        <Login history={historyMock} />
      </AuthContext.Provider>
    </MemoryRouter>
  );
  expect(historyMock.push).toHaveBeenCalled();
  expect(historyMock.push).toHaveBeenCalledWith("/");
});

it("renders the error message when failing to log in", async () => {
  const { getByText, queryByTestId, getByPlaceholderText } = render(
    <MemoryRouter>
      <AuthContext.Provider value={{ setToken: setTokenMock }}>
        <Login history={historyMock} />
      </AuthContext.Provider>
    </MemoryRouter>
  );
  fireEvent.change(getByPlaceholderText("username"), {
    target: {
      value: "skippednote"
    }
  });
  fireEvent.change(getByPlaceholderText("password"), {
    target: {
      value: "skippednote"
    }
  });
  fireEvent.click(getByText("Login"));
  expect(setTokenMock).toBeCalledWith(
    `Basic ${btoa("skippednote:skippednote")}`
  );
  expect(queryByTestId("status-success")).toBeTruthy();
});

it("renders the success message when logged in", async () => {
  const { getByText, queryByTestId, getByPlaceholderText } = render(
    <MemoryRouter>
      <AuthContext.Provider value={{ setToken: setTokenMock }}>
        <Login history={historyMock} />
      </AuthContext.Provider>
    </MemoryRouter>
  );

  fireEvent.click(getByText("Login"));
  expect(setTokenMock).not.toHaveBeenCalled();
  expect(queryByTestId("status-success")).toBeFalsy();
});
