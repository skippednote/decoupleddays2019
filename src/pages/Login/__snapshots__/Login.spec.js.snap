// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div>
  <div>
    Loading component...
  </div>
</div>
`;

exports[`renders the component 2`] = `
<div>
  <form>
    <div
      class="form-item"
    >
      <input
        name="username"
        placeholder="username"
        type="text"
      />
    </div>
    <div
      class="form-item"
    >
      <input
        name="password"
        placeholder="password"
        type="password"
      />
    </div>
    <button
      class="button"
      type="submit"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 3h4a2 2 0 0 1 2 2v14a2 2 0 0 1-2 2h-4"
        />
        <polyline
          points="10 17 15 12 10 7"
        />
        <line
          x1="15"
          x2="3"
          y1="12"
          y2="12"
        />
      </svg>
          Login
    </button>
  </form>
</div>
`;
